### PR TITLE
[Processing][TilesXYZ] Multithreaded Tile generation with ThreadPoolExecutor

### PR DESCRIPTION
**Pull request by isghj5**
Rendering tiles is changed from single thread to multithread using python standard library ThreadPoolExecutor (TPE) class, resulting in a large performance boost for multithreaded CPUs.

Most of the looped code in TilesXYZAlgorithmBase::generate() was relocated into a separate function (per metatile), then all metatiles are mapped to this function by the TPE, which handles thread management for us. To prevent the thread conflicts on the settings, settings were duplicated per thread. 

~~There is one additional addition I think would be good to add: The user interface should have an option to set how many threads are created for the rendering job, right now the code uses all cores available to it(multithreading::cpu_count()) but I don't know how to add this to the user interface.~~ done

This is a re-attempt of a previous pull-request: https://github.com/qgis/QGIS/pull/30473

## Checklist

- [X] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [X] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [X] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

Not sure if this counts as a feature..? It's more a change than an addition I think.

This isn't core, right? It's processing, and I don't know how to make unit tests or what new tests would need to be created

### CODE REVIEW

**Comment by saberraz** [comment]
I agree, having number of cores will be good. But, meanwhile, maybe you can use the number of core user has set to render maps?

**Comment by isghj5** [comment]
> I agree, having number of cores will be good. But, meanwhile, maybe you can use the number of core user has set to render maps?

Assuming you mean Options->Rendering->Rendering Behavior->"Max Cores to use", I'm not sure yet how to get this value. I think it's in QGSServerSettings, however

`from qgis.sever import QGSServerSettings
...
maxThreads = QGSServerSettings().maxThreads()`
Gets me a ModuleNotFoundError: No module named 'qgis.server', even though several other files can find qgis.server, TilesXYZ.py cannot...? Not sure what's going on here, might be a quirk of my machine.

**Comment by roya0045** [comment]
Shouldn't it be in the general settings and not exclusive to the server? I assume that users should also benefit from this as the settings are also located in the rendering/rendering behavior.

**Comment by isghj5** [comment]
Found a function on QgsApplication that returns the core restriction.

I still think there should be an option for the user to set the core count just for a heavy job like rendering tiles on all cores, in case they want to lower cores for just that one task and not have to leave the window, go back to settings, change it, then revert after the task is over, but this is a better default.

File: python/plugins/processing/algs/qgis/TilesXYZ.py [hunk 1]
```diff
@@ -180,102 +184,132 @@ def prepareAlgorithm(self, parameters, context, feedback):
...
+            # Haven't found a better way to break than to make all the new threads return instantly
+
+        if "Dummy" in threading.current_thread().name: # single thread testing
+            threadSpecificSettings = list(self.settingsDictionary.values())[0]
+        else:
+            try:
+                # guess we're not the only one using TPE in QGIS, cannot assume 0_#, it's sometimes 3 or 4_#
+                threadSpecificSettings = self.settingsDictionary[threading.current_thread().name[-1]] # last number only
+            except:
+                print("Exception! our threads don't match with our settings! ")
```
Comment (m-kuhn) [inline python/plugins/processing/algs/qgis/TilesXYZ.py]: Use the message log or other means for debug messages. Or if the exception really shouldn't happen and you want someone to debug that if it happens, just remove the try-except to raise the exception including stack trace for debugging reasons.

File: python/plugins/processing/algs/qgis/TilesXYZ.py [hunk 2]
```diff
@@ -180,102 +184,132 @@ def prepareAlgorithm(self, parameters, context, feedback):
...
-                    tile_img = image.copy(tile_width * r, tile_height * c, tile_width, tile_height)
-                    writer.write_tile(tile, tile_img)
-
-                progress += 1
-                feedback.setProgress(100 * (progress / metatiles_count))
+        self.writer = writer
+
+        self.progressThreadLock = threading.Lock()
+        if self.maxThreads > 1:
+            feedback.pushConsoleInfo('Using %s CPU Threads:' % self.maxThreads)
```
Comment (m-kuhn) [inline python/plugins/processing/algs/qgis/TilesXYZ.py]:
```suggestion
            feedback.pushConsoleInfo(self.tr('Using {max_threads} CPU Threads:').format(max_threads=self.maxThreads))
```

File: python/plugins/processing/algs/qgis/TilesXYZ.py [hunk 3]
```diff
@@ -180,102 +184,132 @@ def prepareAlgorithm(self, parameters, context, feedback):
...
+            feedback.pushConsoleInfo('Pushing all tiles at once: %s tiles.' % len(allMetatiles))
```
Comment (m-kuhn) [inline python/plugins/processing/algs/qgis/TilesXYZ.py]: As above, translate and use format

File: python/plugins/processing/algs/qgis/TilesXYZ.py [hunk 4]
```diff
@@ -180,102 +184,132 @@ def prepareAlgorithm(self, parameters, context, feedback):
...
+            with ThreadPoolExecutor(max_workers=self.maxThreads) as threadPool:
+                threadPool.map(self.renderSingleMetatile, allMetatiles)
+        else:
+            feedback.pushConsoleInfo('Using 1 CPU Thread:')
```
Comment (m-kuhn) [inline python/plugins/processing/algs/qgis/TilesXYZ.py]: translate everywhere, actually :)

**Comment by isghj5** [comment]
Am I correct in interpreting that the reason I cannot find a UI/Layout for this alg is because QgsAlgorithm creates one automatically from the parameter list? 

If this is true, and I wanted to make a new UI element for CPU core to use, where would be the best place to put it in the current layout?

If a layout file was created, could the layout be changed directly? I think it's a bit weird that each small item like zoom_min and zoom_max need to be on separate lines taking up more horizontal space than is really needed...

**Comment by nyalldawson** [comment]
@lsghj5

Processing interfaces are automatically constructed, based on the parameters and output for the algorithm.

I'd suggest this new setting is instead exposed as a processing configuration option, i.e. is set from the Options - Processing tab, and is then usable for other algorithms too.

**Comment by isghj5** [comment]
> I'd suggest this new setting is instead exposed as a processing configuration option, i.e. is set from the Options - Processing tab, and is then usable for other algorithms too.

In Options->Processing->General? Not sure how to add a UI element to set a max core count there, myself.

I still feel for a CPU heavy task like this the user might want to lower their threads for just one task and not have to go searching in a different menu for the option, changing it for one job then changing it back, but we can use the processing setting as a default when it gets added.

**Comment by nyalldawson** [comment]
Well - I'm -1 about exposing an option for this in the algorithm itself. It's cluttering the algorithm's interface, making it more complex for users, and will need to be duplicated across many other algorithms in future.

Ideally, this would be added to QgsProcessingContext as a per-algorithm option, and then we'd have a button in the algorithm dialog to open an "algorithm settings" dialog, with settings like this, a map-extent filter option, options to override the default invalid geometry handling, set output CRSes, etc. 

But that's not a trivial change. I'd love to mentor it if you want to take it on though!

Otherwise, I'd rather a temporary band-aid solution (such as a QSettings key not exposed in the UI, or a Processing configuration key) vs an explicit algorithm parameter which we'll be forced to keep... forever.

**Comment by isghj5** [comment]
> Otherwise, I'd rather a temporary band-aid solution (such as a QSettings key not exposed in the UI, or a Processing configuration key) vs an explicit algorithm parameter which we'll be forced to keep... forever.

Wait wait, I meant adding a parameter to TilesXYZAlgorithmBase, not to QgsAlgorithm. I just wrote a working parameter and it's < 15 lines of code that looks really easy to remove later if so desired.

Is there some project directive I missed saying no UI elements are every allowed to be removed after they are added, as part of design policy or something? I mean, removing UI elements isn't great design, but your suggested setting would replace it, only the UI element would be removed opening up space, the functionality would be replaced not removed completely.

**Comment by nyalldawson** [comment]
> Wait wait, I meant adding a parameter to TilesXYZAlgorithmBase, not to QgsAlgorithm. I just wrote a working parameter and it's < 15 lines of code that looks really easy to remove later if so desired

Right - but that parameter effectively becomes part of stable API, and can never be removed in future when we get a proper way of handling this use case.

> Is there some project directive I missed saying no UI elements are ever allowed to be removed after they are added, as part of design policy or something? 

No -- the issue is about stable API. Processing algorithm's ARE bound by the stable API contract, and adding parameters to an algorithm adds this parameter to the algorithm's API. So we can't remove these in future without breaking the stable API contract.

**Comment by isghj5** [comment]
So, can't make a UI element because it's auto generated from parameters which are used for API too...

Well, how do we add that setting to the general settings then?

**Comment by nyalldawson** [comment]
> Well, how do we add that setting to the general settings then?

Have a look around

https://github.com/qgis/QGIS/blob/master/python/plugins/processing/core/ProcessingConfig.py#L67

**Comment by mash-graz** [comment]
unfortunately i have to report some serious issues related to this changes. -- see  #32471

**Pull request merged**